sycl/test-e2e/syclcompat/kernel/kernel_function_pointer.cpp

```diff
@@ -0,0 +1,244 @@
+// ====------ kernel_function_pointer.cpp---------- -*- C++ -*
+// ----------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//
+// ===---------------------------------------------------------------------===//
+
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+
+#include <iostream>
+
+#include <sycl/detail/core.hpp>
+
+#include <syclcompat/defs.hpp>
+#include <syclcompat/device.hpp>
+#include <syclcompat/kernel.hpp>
+#include <syclcompat/memory.hpp>
+#include <syclcompat/util.hpp>
+
+void vectorAdd(const int *A, int *B, int *C, int N,
+               const sycl::nd_item<3> &item_ct1) {
+  int i = item_ct1.get_group(2) * item_ct1.get_local_range(2) +
+          item_ct1.get_local_id(2);
+  if (i < N) {
+    C[i] = A[i] + B[i];
+  }
+}
+
+// SYCL kernel wrapper.
+void vectorAdd_wrapper(const int *A, int *B, int *C, int N) {
+  sycl::queue queue = *syclcompat::kernel_launcher::_que;
+  unsigned int localMemSize = syclcompat::kernel_launcher::_local_mem_size;
+  sycl::nd_range<3> nr = syclcompat::kernel_launcher::_nr;
+
+  queue.parallel_for(
+      nr, [=](sycl::nd_item<3> item_ct1) { vectorAdd(A, B, C, N, item_ct1); });
+}
+
+template <typename T>
+void vectorTemplateAdd(const T *A, T *B, T *C, int N,
+                       const sycl::nd_item<3> &item_ct1) {
+  int i = item_ct1.get_group(2) * item_ct1.get_local_range(2) +
+          item_ct1.get_local_id(2);
+  if (i < N) {
+    C[i] = A[i] + B[i];
+  }
+}
+
+// SYCL kernel wrapper.
+template <typename T>
+void vectorTemplateAdd_wrapper(const T *A, T *B, T *C, int N) {
+  sycl::queue queue = *syclcompat::kernel_launcher::_que;
+  unsigned int localMemSize = syclcompat::kernel_launcher::_local_mem_size;
+  sycl::nd_range<3> nr = syclcompat::kernel_launcher::_nr;
+
+  queue.parallel_for(nr, [=](sycl::nd_item<3> item_ct1) {
+    vectorTemplateAdd<T>(A, B, C, N, item_ct1);
+  });
+}
+
+void hostCallback(void *userData) {
+  const char *msg = static_cast<const char *>(userData);
+  std::cout << "Host callback executed. Message: " << msg << std::endl;
+}
+
+template <typename T> using fpt = void (*)(const T *, T *, T *, int);
+
+void test_kernel_launch() {
+  sycl::queue q_ct1 = syclcompat::get_default_queue();
+  int N = 10;
+  size_t size = N * sizeof(int);
+
+  int *h_A = new int[N];
+  int *h_B = new int[N];
+  int *h_C = new int[N];
+
+  for (int i = 0; i < N; ++i) {
+    h_A[i] = static_cast<int>(i);
+    h_B[i] = static_cast<int>(i * 2);
+  }
+
+  int *d_A, *d_B, *d_C;
+  d_A = (int *)sycl::malloc_device(size, q_ct1);
+  d_B = (int *)sycl::malloc_device(size, q_ct1);
+  d_C = (int *)sycl::malloc_device(size, q_ct1);
+
+  q_ct1.memcpy(d_A, h_A, size).wait();
+  q_ct1.memcpy(d_B, h_B, size).wait();
+
+  fpt<int> fp = syclcompat::wrapper_register(vectorAdd_wrapper).get();
+
+  void *kernel_func =
```

Comment:
kernel_func is unused here. Maybe it was meant to be passed to `syclcompat::kernel_launcher::launch` in place of fp? although it seems to be getting cast correctly in that call.